Plotting Fractals

MichaelRipa opened the pull request:
Hi NumPy Devs, 

As proposed in #92 , this tutorial demonstrates how to plot fractal visuals using NumPy and matplotlib. It also makes use of the mpl_toolkits library (to fix formatting issues with the colorbars). All feedback is welcome, please don't hesitate to request changes of any kind! 

Thanks again for your time!  

Michael

rossbar commented:
Thanks for this very nice tutorial @MichaelRipa ! I took the liberty of pushing up a few changes related to the workflow so that we could get the CI running correctly. The two most relevant changes are:
 - I added your new tutorial to the `toctree` in `site/applications.md` - this guarantees that the new tutorial will be included in the generated html site.
 - I converted the notebook itself from `.ipynb` to the [text-based notebook format](https://myst-nb.readthedocs.io/en/latest/use/markdown.html) used in this repo. This makes writing and especially reviewing notebooks much more straightforward. Note that if you have `jupytext` installed (see `environments.yml` or `requirements.txt`) the notebook will still open in the browser the way you are used to with an `.ipynb` file, so this shouldn't impact your workflow at all --- please let us know if you have problems or questions though!

With the CI passing, you can now view the rendered notebook from the `build-docs artifact` link in the CI checks above. Here's a link to the new tutorial for convenience: https://266-248354526-gh.circle-artifacts.com/0/site/_build/html/content/tutorial-plotting-fractals.html

I've only had a chance to skim the content of the tutorial itself so far, but at first glance it looks excellent! I will aim to take a closer look as soon as I have time.

mattip commented:
I like the progression of the tutorial from basics to beautiful fractals. I think it would be helpful to use links in the body of the tutorial to relevant NumPy, matplotlib documentation and wiki articles.

MichaelRipa commented:
@rossbar Thanks for taking the time to fix up the submission errors and converting the notebook type, as well as for your feedback! My local copy of the notebook is now paired with MyST and so any future submissions on my end should be consistant with the repository. 

@mattip Thanks your feedback, that's a great suggestion! I can start working on implementing this change, I will start by  adding links for Universal Functions, meshgrid, 3d scatterplots, Boolean indexing and imshow as they would be the easiest to add in. I also will add wiki links to the prerequisites  I mentioned in the "What you'll need" section. Please let me know if there are other things that would be good to link into the tutorial.

rossbar commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,558 @@
+    return diverge_len
+```
+
+Now we can experiment with some different polynomials.
+
+For $p(z) = z^8 + 15 z^4 - 16$, we have 
+
+
+$\frac{dp}{dz} = 8z^7 + 60z^3$.
+
+Putting the polynomials to code and calling the function gives us:
+
+```{code-cell} ipython3
+def poly(z):
+    return np.power(z,8) + 15* np.power(z,4) -16
+
+def derivative(z):
+    return 8* np.power(z,7) + 60 * np.power(z,3)
+```
+
+```{code-cell} ipython3
+output = newton_fractal(mesh,poly,derivative,num_iter=15,r=2)
+kwargs = {'title':'f(z) = z - \dfrac{(z^8 + 15z^4 - 16)}{(8z^7 + 60z^3)}', 'cmap':'copper'}
+
+plot_fractal(output,**kwargs)
+```
````
This is a nice opportunity to highlight [NumPy's polynomial package](https://numpy.org/doc/stable/reference/routines.polynomials.html). Note that polynomial objects should render w/ TeX in Jupyter notebooks.

````suggestion
Now we can experiment with some different polynomials.
For example, let's try a higher-degree polynomial:

```{code-cell} ipython3
p = np.polynomial.Polynomial([-16, 0, 0, 0, 15, 0, 0, 0, 1])
p
```

which has the derivative:

```{code-cell} ipython3
p.deriv()
```

```{code-cell} ipython3
output = newton_fractal(mesh,p ,p.deriv(), num_iter=15,r=2)
kwargs = {'title':'f(z) = z - \dfrac{(z^8 + 15z^4 - 16)}{(8z^7 + 60z^3)}', 'cmap':'copper'}

plot_fractal(output,**kwargs)
```
````

rossbar reviewed (changes requested):
Very nice tutorial @MichaelRipa !

I took a first pass through the code and went ahead and made a few minor changes along with some suggestions. I mostly focused on the code cells rather than the text.

I did notice one high-level terminology issue, though as I mention in the comments, I'm not an expert so maybe my understanding is flawed: I think some of the sets that are referred to in this tutorial as "generalized Mandelbrot sets" or similar are more accurately described as [Julia sets](https://en.wikipedia.org/wiki/Julia_set). If I'm missing something on this point LMK, otherwise I think it'd be good to tighten up the wording in the Mandelbrot text.

Once these are addressed I'd like to do one last pass as more of a copy-edit, but IMO this is already in great shape. Thanks again!

MichaelRipa commented:
@rossbar Really appreciate you taking the time to go through the tutorial and provide feedback; your updates are excellent and the comments are a huge help! I will begin going through the updates and comments over the next few days and begin incorporating the suggestions you provided.

Again, a huge thanks for your updates, they really help improve the readability and I have learned a lot from them!

rossbar commented:
Just a quick note: the failing CI had nothing to do with your tutorial @MichaelRipa. I pushed up a temporary fix in 5d81214 just to keep the review process flowing.

MichaelRipa commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,558 @@
+With that out of the way, we can go about plotting our first fractal! We will use the [**imshow**](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.imshow.html) function to create a colour-coded visualization of the tallies. 
+
+```{code-cell} ipython3
+x,y = np.meshgrid(np.arange(-2,2,0.01),np.arange(-2,2,0.01))
````
Good catch, I'll replace all instances of `np.arange` in my next commit to maintain consistency.

MichaelRipa commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,558 @@
+
+# Mandelbrot 
+
+What we just explored was an example of a fractal found in the Mandelbrot set. We can write a more generalized version of the above function to gain access to a wider range of fractals.
```
Really glad you brought this up, I learned a lot of what I know about fractals by writing this tutorial, somehow missed this connection! Before my next commit, I will do some additional reading on Julia sets and will modify the tutorial to briefly explain what it is. Luckily this should be relatively easy to update since there are only two sections with the "general Mandelbrot function" where the first section actually is plotting the Julia set. Thanks for bringing this up, it would have been a shame to have had this slip past revision!

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
++++
+
+Fractals are beautiful, compelling mathematical forms that can be oftentimes created from a relatively simple set of instructions. In nature they can be found in various places, such as coastlines, seashells, and ferns, and even were used in creating certain types of antennas. The mathematical idea of fractals was known for quite some time, but they really began to be truly appreciated in the 197 ...
```
Can we add a link to Mandelbrot's wikipedia page? 

```suggestion
Fractals are beautiful, compelling mathematical forms that can be oftentimes created from a relatively simple set of instructions. In nature they can be found in various places, such as coastlines, seashells, and ferns, and even were used in creating certain types of antennas. The mathematical idea of fractals was known for quite some time, but they really began to be truly appreciated in the 1970's as advancements in computer graphics and some accidental discoveries lead researchers like [Benoît Mandelbrot](https://en.wikipedia.org/wiki/Benoit_Mandelbrot) to stumble upon the truly mystifying visualizations that fractals possess. 
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+## What you'll need
+
+- [Matplotlib](https://matplotlib.org/)
+- make_axis_locatable function from mpl_toolkits API
```
```suggestion
- `make_axis_locatable` function from mpl_toolkits API
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+To gain some intuition for the behaviour of the function, we can try plugging in some different values.
+
+For z = 0, we would expect to get -1
```
nit - just so that formatting is consistent:
```suggestion
For $z = 0$, we would expect to get $-1$:
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+$f(z) = z^2 -1 $ 
+
+where z is a complex number (i.e of the form $a + bi$ )
```
```suggestion
where `z` is a complex number (i.e of the form $a + bi$ )
```

melissawm commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,553 @@
+f(0)
+```
+
+Some values grow, some values shrink, some don't experience much change.
+
+```{code-cell} ipython3
+z = [4, 1-0.2j, 1.6]
+f(z)
+```
````
I think it would be useful to note quickly that this is the same as applying the function to the three values separately; otherwise it might be hard to understand what we should be noticing in the values here.

MichaelRipa replied:
Good insight, I will incorporate this into my next update.

melissawm commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,553 @@
+    return np.square(z) - 1  
+```
+
+Note that the square function we used is an example of a **[NumPy Universal Function](https://numpy.org/doc/stable/reference/ufuncs.html)**; we will come back to the significance of this decision shortly. 
````
We have [a PR up](https://github.com/numpy/numpy/pull/19516) to update this doc - so we might need to change this link in the future. Also - @rossbar do you think it would be better to use intersphinx for the tutorials? This had escaped me before but now I'm starting to think it would be the best solution.

rossbar replied:
> Also - @rossbar do you think it would be better to use intersphinx for the tutorials? This had escaped me before but now I'm starting to think it would be the best solution.

The reason we decided not to use `intersphinx` links in the tutorials themselves is because they will only render in the html output, not in the notebooks (e.g. for binder users). I haven't been paying super close attention to when/whether the browser-based interfaces will support MyST syntax, but there seems to be promising work in support of that general direction, e.g. executablebooks/meta#449. 

I don't think we have any data on the number of binder users (do we?) so I think it's probably best to stick with commonmark compliant markdown for now so that the links work for all users. If/when Jupyter supports MyST in the browser, it'd be a great mini-sprint activity to go through and "myst-ify" all the notebooks to take advantage of all the awesome additional features :)

Of course, using intersphinx to link **to** the tutorials (e.g. from the official numpy docs) is still possible & encouraged (e.g. once this is merged it might be nice to add a link to the Mandelbrot example in the Quickstart tutorial!)

melissawm commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,553 @@
+ax.set_title('One Iteration: $ f(z) = z^2 - 1$');
+```
+
+This gives us a rough idea of what one iteration of the function does. Certain areas (notably in the areas closest to (0,0i)) remain rather small while other areas grow quite considerably. Note that we lose information about the output by taking the absolute value, but it is the only way for us to be able to make a plot.
````
```suggestion
This gives us a rough idea of what one iteration of the function does. Certain areas (notably in the areas closest to $(0,0i)$) remain rather small while other areas grow quite considerably. Note that we lose information about the output by taking the absolute value, but it is the only way for us to be able to make a plot.
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+    index = np.arange(0,10)
+
+
```
```suggestion
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+        diverge_len[np.abs(z) < radius] += 1
+        z[np.abs(z) < radius] = f(z[np.abs(z)< radius])   
+        
+        
```
```suggestion
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+The behaviour of this function may look confusing at first glance, so it will help to explain some of the notation.
+
+Our goal is to iterate over each value in the mesh and to tally the number of iterations before the value diverges. Since some values will diverge quicker than others, we need a procedure that only iterates over values that have an absolute value that is sufficiently small enough. We also want to stop tallying values once they surpass the radius. For this, we can use **[Boolean Indexing](https:// ...
```
```suggestion
Our goal is to iterate over each value in the mesh and to tally the number of iterations before the value diverges. Since some values will diverge quicker than others, we need a procedure that only iterates over values that have an absolute value that is sufficiently small enough. We also want to stop tallying values once they surpass the radius. For this, we can use **[Boolean Indexing](https://numpy.org/devdocs/reference/arrays.indexing.html#boolean-array-indexing)**, a NumPy feature that when paired with Universal Functions is unbeatable. Boolean Indexing allows for operations to be performed conditionally on a NumPy array without having to resort to looping over and checking for each array value individually. 
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+We also can start experimenting with different values of c. It can be surprising how much influence it has on the shape of the fractal.
+
+For example, setting $c = \frac{\pi}{10}$ gives us a very elegent cloud shape, while setting c = $-\frac{3}{4} + 0.4i$ yields a completely different pattern.
```
```suggestion
For example, setting $c = \frac{\pi}{10}$ gives us a very elegant cloud shape, while setting c = $-\frac{3}{4} + 0.4i$ yields a completely different pattern.
```

melissawm commented on content/tutorial-plotting-fractals.md:
````diff
@@ -0,0 +1,553 @@
+plot_fractal(output,**kwargs);
+```
+
+What happens if we compose our defined function inside of a sin function?
````
```suggestion
What happens if we compose our defined function inside of a sine function?
```

melissawm commented on content/tutorial-plotting-fractals.md:
```diff
@@ -0,0 +1,553 @@
+
+## Further reading 
+
+More information on the theory behind fractals - https://en.wikipedia.org/wiki/Fractal 
+
+A list of different fractals - https://en.wikipedia.org/wiki/List_of_fractals_by_Hausdorff_dimension
```
```suggestion
- [More information on the theory behind fractals](https://en.wikipedia.org/wiki/Fractal)
- [A list of different fractals](https://en.wikipedia.org/wiki/List_of_fractals_by_Hausdorff_dimension)
```

melissawm reviewed (changes requested):
This is such a cool tutorial, thank you @MichaelRipa !  

I left a few very small comments but overall this looks really good. If you can, would you mind running a PEP8 checker on the code when you update the tutorial?

MichaelRipa commented:
Thank you @melissawm for your review of the tutorial, your feedback and effort is very much appreciated! I am currently not somewhere with a good internet connection, but I will commit the suggestions early next week and should have my update ready shortly after! (which I will run a PEP8 checker on)

rossbar commented:
> which I will run a PEP8 checker on

Just a quick tip: you can use `jupytext` to run a PEP8 checker, e.g. `jupytext tutorial-plotting-fractals.md --check flake8`. You may want to suppress the E501 warnings though because you will get a lot errors from the markdown cells where text tends to be longer than 80chars.

An (IMO better) alternative is to use `jupytext` to *apply* formatting rules to the whole file automatically, e.g. with `black`: `jupytext tutorial-plotting-fractals.md --pipe black`. Instead of only warning about linting errors, this will automatically fix them, though `black` is far more opinionated than a pure pep8 linter would be.

MichaelRipa commented:
Thanks again @rossbar and @melissawm for your updates, they really make a difference in the quality of the tutorial! I finished up my updates for now, I believe that I addressed all of the issues brought up earlier but do let me know if I missed anything. 

Here is a quick summary of the major changes: 

- I added a very short new section that discusses and plots the Mandelbrot set, as it turned out that I had not plotted it at all in the previous update. I also gave concise definitions for the filled-in Julia set, Julia set and Mandelbrot set, as my previous update considered everything as a "Mandelbrot fractal" which was wrong.  This resulted in me renaming certain sections and functions to keep consistant with the update.

- I ran a PEP8 check and fixed a large majority of the errors. I left a few that I was unsure about:
  - W605 (Shows up from using '\\' in the title of certain plots)
  - E703 (Using semi-colons to suppress output) 
  - E731 (Using a lambda function instead of def)
  - E7501 
 
- I briefly mentioned the [Polynomial class](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.html) and linked to its documentation.

- I added wiki links for the Mandelbrot set, Julia set and for Newton fractals in the "Further reading" section.

Thanks for being patient with this one, it took me longer than expected to finish. As always, don't hesitate to provide any feedback or suggestions to the updates, all of your help has been very much appreciated thus far!

melissawm reviewed (approved):
I am happy with the tutorial, looks great - thanks again @MichaelRipa for the awesome content!

rossbar commented:
This LGTM, thanks @MichaelRipa for the idea and hard work & @melissawm for the nice suggestions! Let's put this in and iterate on any improvements/ideas in follow-up PRs.

MichaelRipa commented:
I just wanted to thank everyone again for all the help and encouragement. This was my first experience contributing to open source and I couldn't have gotten a better introduction to the open-source community then I did here! Hope to get to work with you all again on future projects, see you in the follow-up PRs!
